compiler/src/dotty/tools/backend/sjs/JUnitBootstrappers.scala

```diff
@@ -0,0 +1,341 @@
+    def BeforeAnnotClass(implicit ctx: Context): ClassSymbol = BeforeAnnotType.symbol.asClass
+
+    lazy val AfterAnnotType: TypeRef = ctx.requiredClassRef("org.junit.After")
+    def AfterAnnotClass(implicit ctx: Context): ClassSymbol = AfterAnnotType.symbol.asClass
+
+    lazy val BeforeClassAnnotType: TypeRef = ctx.requiredClassRef("org.junit.BeforeClass")
+    def BeforeClassAnnotClass(implicit ctx: Context): ClassSymbol = BeforeClassAnnotType.symbol.asClass
+
+    lazy val AfterClassAnnotType: TypeRef = ctx.requiredClassRef("org.junit.AfterClass")
+    def AfterClassAnnotClass(implicit ctx: Context): ClassSymbol = AfterClassAnnotType.symbol.asClass
+
+    lazy val IgnoreAnnotType: TypeRef = ctx.requiredClassRef("org.junit.Ignore")
+    def IgnoreAnnotClass(implicit ctx: Context): ClassSymbol = IgnoreAnnotType.symbol.asClass
+
+    lazy val BootstrapperType: TypeRef = ctx.requiredClassRef("org.scalajs.junit.Bootstrapper")
+
+    lazy val TestMetadataType: TypeRef = ctx.requiredClassRef("org.scalajs.junit.TestMetadata")
+
+    lazy val NoSuchMethodExceptionType: TypeRef = ctx.requiredClassRef("java.lang.NoSuchMethodException")
+
+    lazy val FutureType: TypeRef = ctx.requiredClassRef("scala.concurrent.Future")
+    def FutureClass(implicit ctx: Context): ClassSymbol = FutureType.symbol.asClass
+
+    private lazy val FutureModule_successfulR = ctx.requiredModule("scala.concurrent.Future").requiredMethodRef("successful")
+    def FutureModule_successful(implicit ctx: Context): Symbol = FutureModule_successfulR.symbol
+
+    private lazy val SuccessModule_applyR = ctx.requiredModule("scala.util.Success").requiredMethodRef(nme.apply)
+    def SuccessModule_apply(implicit ctx: Context): Symbol = SuccessModule_applyR.symbol
+  }
+
+  // The actual transform -------------------------------
+
+  override def transformPackageDef(tree: PackageDef)(implicit ctx: Context): Tree = {
+    // TODO Can we cache this better?
+    implicit val mydefn: MyDefinitions = new MyDefinitions()
+
+    @tailrec
+    def hasTests(sym: ClassSymbol): Boolean = {
+      sym.info.decls.exists(m => m.is(Method) && m.hasAnnotation(mydefn.TestAnnotClass)) ||
+      sym.superClass.exists && hasTests(sym.superClass.asClass)
+    }
+
+    def isTestClass(sym: Symbol): Boolean = {
+      sym.isClass &&
+      !sym.is(ModuleClass | Abstract | Trait) &&
+      hasTests(sym.asClass)
+    }
+
+    val bootstrappers = tree.stats.collect {
+      case clDef: TypeDef if isTestClass(clDef.symbol) =>
+        genBootstrapper(clDef.symbol.asClass)
+    }
+
+    if (bootstrappers.isEmpty) tree
+    else cpy.PackageDef(tree)(tree.pid, tree.stats ::: bootstrappers)
+  }
+
+  private def genBootstrapper(testClass: ClassSymbol)(
+      implicit ctx: Context, mydefn: MyDefinitions): TypeDef = {
+
+    val owner = testClass.owner
+    val moduleSym = ctx.newCompleteModuleSymbol(owner,
+        (testClass.name ++ "$scalajs$junit$bootstrapper").toTermName,
```

Comment:
If possible, using a semantic name would be better, see e.g. https://github.com/lampepfl/dotty/blob/fbc9d0a4144cc0330422c94497130edf7aa541e3/compiler/src/dotty/tools/dotc/core/NameKinds.scala#L369

Reply:
I'd say that the same reasoning apply to the dollar we add to module classes, and yet we represent them with a namekind anyway: https://github.com/lampepfl/dotty/blob/fbc9d0a4144cc0330422c94497130edf7aa541e3/compiler/src/dotty/tools/dotc/core/NameKinds.scala#L367. It just makes it easier to find the original name at a later point if needed, but it doesn't change what we get in the bytecode.

Reply:
Hum, OK. I've attempted to do it in the latest commit. But I get the following exception:
```
[info] exception occurred while compiling /localhome/doeraene/projects/dotty/sandbox/scalajs/test/HelloTest.scala
java.lang.AssertionError: assertion failed: module naming inconsistency: HelloTest[ModuleClass][Suffix $scalajs$junit$bootstrapper]/T while compiling /localhome/doeraene/projects/dotty/sandbox/scalajs/test/HelloTest.scala
[error] ## Exception when compiling 1 sources to /localhome/doeraene/projects/dotty/sandbox/scalajs/../out/bootstrap/sjsSandbox/scala-0.15/test-classes
[error] assertion failed: module naming inconsistency: HelloTest[ModuleClass][Suffix $scalajs$junit$bootstrapper]/T
[error] dotty.DottyPredef$.assertFail(DottyPredef.scala:16)
[error] dotty.tools.dotc.core.SymDenotations$ClassDenotation.<init>(SymDenotations.scala:1450)
[error] dotty.tools.dotc.core.SymDenotations.SymDenotation(SymDenotations.scala:42)
[error] dotty.tools.dotc.core.Contexts$Context.SymDenotation(Contexts.scala:71)
[error] dotty.tools.dotc.core.Symbols.newModuleSymbol(Symbols.scala:156)
[error] dotty.tools.dotc.core.Contexts$Context.newModuleSymbol(Contexts.scala:71)
[error] dotty.tools.dotc.core.Symbols.newCompleteModuleSymbol(Symbols.scala:185)
[error] dotty.tools.dotc.core.Contexts$Context.newCompleteModuleSymbol(Contexts.scala:71)
[error] dotty.tools.backend.sjs.JUnitBootstrappers.dotty$tools$backend$sjs$JUnitBootstrappers$$genBootstrapper(JUnitBootstrappers.scala:155)
```
and I don't know how to go about dealing with it at the moment.

Reply:
The name of an object should be a ModuleClassName, so you need to wrap your ScalaJSJUnitBoostrapperName into a ModuleClassName (and then it'll get a $ at the end which is hopefully what you want ?)

Reply:
But the name I need here is a `TermName` which is the name of the *module*, not its module class. If I try to wrap it in a `ModuleClassName` I get the exact same exception...

Reply:
Oh right, it seems that the logic of NameKind is a bit confused between the concept of ordering NameKinds by tag and the concept that certain NameKinds can define a new name. Maybe forget about semantic names for now :)

Reply:
After talking offline, I've reverted to using a non-semantic name. We could possibly revisit this in the future.